File: Search_2D_Array.py
# // Time Complexity : O(logN)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this :


# // Your code here along with comments explaining your approach


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if (matrix == None or len(matrix)==0):
            return False
        m = len(matrix) # number of rows
        n = len(matrix[0]) # number of columns
        low = 0
        high = m*n -1
        while(low<=high):
            mid = low + (high -low)//2
            
            row = int(mid/n)
            col = int(mid%n)
            
            if matrix[row][col] == target:
                return True
            elif matrix[row][col] > target:
                high = mid -1 
            elif matrix[row][col] < target:
                low = mid +1
                
        return False
            
            
            
# Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
# Output: false           
            
        
        

File: search2DArray.java
public class search2DArray {
    /** Time Complexity : O(log(m*n))
     Space Complexity : O(1)
     Did this code successfully run on Leetcode : Yes
     Any problem you faced while coding this : No


     Your code here along with comments explaining your approach in three sentences only **/
    public boolean searchMatrix(int[][] matrix, int target) {
        //These are the rows and columns of the matrix
        int m = matrix.length;
        int n = matrix[0].length;
        //For Binary Search:
        int left = 0;
        int right = m * n - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            int midValue = matrix[mid / n][mid % n];

            if (midValue == target) {
                return true;
            } else if (midValue < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return false;

    }
}
